daf_butler.py

```diff
@@ -0,0 +1,942 @@
+from __future__ import annotations
+
+__all__ = (
+    "QuantumGraphExpression",
+    "QuantumStatus",
+    "DatasetStatus",
+    "QuantumGraph",
+    "WorkspaceButler",
+    "Butler",
+)
+
+import enum
+import uuid
+from abc import abstractmethod, ABC
+from collections.abc import Iterable, Mapping, Sequence, Set
+import json
+from types import EllipsisType
+from typing import Any, ClassVar, Protocol, Self, cast, final
+
+from pydantic import BaseModel
+
+import networkx as nx  # type: ignore[import]
+from lsst.daf.butler import (
+    Config,
+    DataCoordinate,
+    DataId,
+    DatasetRef,
+    DimensionGraph,
+    DimensionUniverse,
+    LimitedButler,
+    Registry,
+)
+from lsst.resources import ResourcePath, ResourcePathExpression
+from lsst.utils.introspection import get_full_type_name
+from lsst.utils.doImport import doImportType
+
+
+class QuantumGraphExpression:
+    """Placeholder for a parsed expression that subsets a quantum graph.
+
+    This is expected to include:
+
+    - the standard set-theory operations (intersection, union, difference,
+      symmetric difference, inversion);
+    - literal sets of quanta and datasets;
+    - range syntax for ancestors and descendants of quanta and datasets (see
+      `PipelineGraphExpression`).
+
+    In both literal sets and range expressions, quanta and datasets could be
+    identified via:
+
+    - UUID literals;
+    - combination of task label or dataset type name and data ID (e.g.
+      ``makeWarp@{tract=9813, patch=42, visit=12}``);
+    - bind aliases for UUIDs, sets of UUIDs, task labels, dataset type names,
+      and data IDs.
+    - the values of the `QuantumStatus` and `DatasetStatus` enums as keywords,
+      representing the set of all quanta or datasets with that status.
+
+    Data IDs keys that are common to all selected nodes can be provided by
+    butler defaults or a separate single data ID, which should be accepted by
+    any method accepting an expression of this type.
+
+    In queries that span multiple collections, dataset and quantum identifiers
+    that use data IDs are resolved in the order they appear in the collection
+    sequence, but range expressions can pick up datasets and quanta that are
+    shadowed.
+    """
+
+
+class QuantumStatus(enum.Enum):
+    """Valid states for a quantum node.
+
+    This is intended only to capture status to the extent necessary to
+    represent state transitions when executing or retrying quanta; full status
+    information that's useful for diagnostic purposes does not belong here.
+    """
+
+    BUILT = enum.auto()
```

Comment:
These values are persisted, maybe select specific values for them instead of `auto` in case someone wants to reorder them later or insert something in the middle?